File: typescript/tests/qualification/operation-worker.mjs
/** Measure installed SDK calls, with generation and oracle work outside call timers. */
import assert from 'node:assert/strict'
import { readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { pathToFileURL } from 'node:url'
const args = Object.fromEntries(Array.from({ length: (process.argv.length - 2) / 2 }, (_, n) =>
  [process.argv[2 + n * 2].replace(/^--/, ''), process.argv[3 + n * 2]]))
/** @type {typeof import('../../src/index.js')} */
const sdk = await import(pathToFileURL(join(args.sdk, 'dist/index.js')).href)
const { PostgresEngine } = await import(pathToFileURL(join(args.sdk, 'dist/engines/postgres.js')).href)
const { ClickHouseEngine } = await import(pathToFileURL(join(args.sdk, 'dist/engines/clickhouse.js')).href)
const config = JSON.parse(readFileSync(args.config, 'utf8'))
const model = sdk.assemble(config.model.entities.map(entity => ({ ...entity,
  fields: entity.fields.map(field => ({ ...field, nullable: field.nullable === true })),
  pii: entity.pii ?? [], residency: entity.residency ?? null })), [], config.model.atomic ?? [], config.model.cost_ceiling ?? null)
const keys = Object.fromEntries(Object.entries(config.public_keys).map(([key, value]) => [key, Buffer.from(value, 'base64')]))
const placement = sdk.loadMap(JSON.parse(readFileSync(join(args['project-dir'], 'active-map.json'), 'utf8')),
  { model, publicKey: keys, requireSignature: true })
const bindingName = placement.groups.WeatherReading.source.engine, binding = config.engines[bindingName]
const Engine = binding.dialect === 'postgres' ? PostgresEngine : ClickHouseEngine
const engine = new Engine(process.env[binding.runtime_dsn_envs[0]])
const recorder = args.telemetry === 'on' ? new sdk.Recorder(model.version) : undefined
const rows = Number(args.rows), batch = Number(args['batch-size']), repetitions = Number(args.samples)
assert.ok(Number.isSafeInteger(rows) && rows > 0 && rows <= 1_000_000)
assert.ok(Number.isSafeInteger(batch) && batch > 0 && batch <= 1000)
assert.ok(Number.isSafeInteger(repetitions) && repetitions > 0 && repetitions <= 10000)
assert.ok(['write', 'read'].includes(args.mode) && ['save', 'save_many'].includes(args.method))
assert.ok(args.method !== 'save' || batch === 1)
const base = sdk.Timestamp.from('2026-09-14T12:00:00.123456Z').epochMicroseconds
function reading(worker, sequence) {
  const cents = 1525 + sequence % 1000
  return { station: `qualification-station-${worker}`,
    at: sdk.Timestamp.fromEpochMicroseconds(base + BigInt(sequence)),
    celsius: `${Math.floor(cents / 100)}.${String(cents % 100).padStart(2, '0')}`,
    humidity: BigInt(30 + sequence % 70),
    id: `${worker.toString(16).padStart(8, '0')}-0000-4000-8000-${sequence.toString(16).padStart(12, '0')}` }
}
const samples = []
/** @type {Record<string, any>} */
const report = { protocol: 1, language: 'typescript', mode: args.mode, method: args.method,
  batch_size: batch, rows, sdk_module: args.sdk, status: 'incomplete', acknowledged: 0,
  samples, telemetry_enabled: recorder !== undefined }
let session, began = process.hrtime.bigint()
try {
  await engine.connect()
  session = await sdk.Session.open(model, placement, { [bindingName]: engine }, { recorder, projectId: config.project_id })
  if (args.mode === 'write') {
    for (let offset = 0; offset < 2; offset++) {
      const values = Array.from({ length: batch }, (_, n) => reading(98, offset * batch + n + 1))
      if (args.method === 'save') await session.save('WeatherReading', values[0])
      else await session.saveMany('WeatherReading', values)
    }
    began = process.hrtime.bigint()
    for (let first = 1; first <= rows; first += batch) {
      const values = Array.from({ length: Math.min(batch, rows - first + 1) }, (_, n) => reading(1, first + n))
      const start = process.hrtime.bigint()
      if (args.method === 'save') await session.save('WeatherReading', values[0])
      else await session.saveMany('WeatherReading', values)
      const duration = Number(process.hrtime.bigint() - start)
      samples.push({ operation: args.method, duration_ns: duration, rows: values.length })
      report.acknowledged += values.length
    }
  } else {
    const where = { station: reading(99, 1).station }
    const cents = 1525n * BigInt(rows) + BigInt(Math.floor(rows / 1000)) * 499500n + BigInt(rows % 1000) * BigInt(rows % 1000 + 1) / 2n
    const total = `${cents / 100n}.${String(cents % 100n).padStart(2, '0')}`
    for (let n = 0; n < 20; n++) assert.deepEqual(await session.get('WeatherReading', { ...where, at: reading(99, 1).at }), reading(99, 1))
    began = process.hrtime.bigint()
    for (let n = 0; n < repetitions; n++) {
      const sequence = 1 + n * 104729 % rows, expected = reading(99, sequence)
      for (const operation of ['get', 'scan', 'count', 'summarize']) {
        const start = process.hrtime.bigint()
        let result
        if (operation === 'get') result = await session.get('WeatherReading', { ...where, at: expected.at })
        else if (operation === 'scan') result = await session.scan('WeatherReading', { where, bounds: { field: 'at', low: expected.at }, limit: 1000 })
        else if (operation === 'count') result = await session.count('WeatherReading', { where })
        else result = await session.summarize('WeatherReading', 'celsius', { where })
        const duration = Number(process.hrtime.bigint() - start)
        let returned = 1
        if (operation === 'get') assert.deepEqual(result, expected)
        else if (operation === 'scan') {
          assert.ok(result !== null && typeof result === 'object' && 'rows' in result)
          returned = Math.min(1000, rows - sequence + 1)
          assert.deepEqual(result.rows, Array.from({ length: returned }, (_, k) => reading(99, sequence + k)))
        } else if (operation === 'count') assert.equal(result, BigInt(rows))
        else {
          assert.ok(result !== null && typeof result === 'object' && 'count' in result && 'nonNullCount' in result && 'total' in result)
          assert.equal(result.count, BigInt(rows)); assert.equal(result.nonNullCount, BigInt(rows)); assert.equal(result.total, total)
        }
        samples.push({ operation, duration_ns: duration, rows: returned })
      }
    }
  }
  report.elapsed_ns = Number(process.hrtime.bigint() - began)
  report.status = 'complete'
} finally {
  try { if (session) await session.close(); await engine.close() }
  finally {
    report.max_rss_kib = process.resourceUsage().maxRSS
    if (recorder) { const window = recorder.roll(); assert.ok(window); report.window = sdk.windowRecord(window, model) }
    writeFileSync(args.output, JSON.stringify(report))
  }
}

File: python/tests/qualification/operation_worker.py
"""Measure installed SDK calls; generate and check synthetic values outside each call timer."""

from __future__ import annotations

import argparse
import base64
import json
import os
import resource
import time
from datetime import timedelta
from decimal import Decimal
from pathlib import Path
from typing import Any

import sde
from sde.engines.clickhouse import ClickHouseEngine
from sde.engines.postgres import PostgresEngine
from sde.testing.loader import model_from_neutral

from .weather_worker import BASE_TIME, reading


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--config", type=Path, required=True)
    parser.add_argument("--project-dir", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--mode", choices=("write", "read"), required=True)
    parser.add_argument("--method", choices=("save", "save_many"), default="save_many")
    parser.add_argument("--rows", type=int, required=True)
    parser.add_argument("--batch-size", type=int, default=100)
    parser.add_argument("--samples", type=int, default=500)
    parser.add_argument("--telemetry", choices=("on", "off"), default="on")
    args = parser.parse_args()
    if (
        not 1 <= args.rows <= 1_000_000
        or not 1 <= args.batch_size <= 1000
        or not 1 <= args.samples <= 10000
    ):
        parser.error("bounded positive rows, batch and samples are required")
    if args.method == "save" and args.batch_size != 1:
        parser.error("save measures exactly one row")
    config = json.loads(args.config.read_bytes())
    model = model_from_neutral(config["model"])
    keys = {name: base64.b64decode(value) for name, value in config["public_keys"].items()}
    placement = sde.load_local_map(
        args.project_dir, model=model, project_id=config["project_id"], public_key=keys
    )
    name = placement.groups["WeatherReading"].source.engine
    binding = config["engines"][name]
    kind: Any = PostgresEngine if binding["dialect"] == "postgres" else ClickHouseEngine
    recorder = sde.Recorder(model.version) if args.telemetry == "on" else None
    samples: list[dict[str, Any]] = []
    report: dict[str, Any] = {
        "protocol": 1,
        "language": "python",
        "mode": args.mode,
        "method": args.method,
        "batch_size": args.batch_size,
        "rows": args.rows,
        "sdk_module": sde.__file__,
        "status": "incomplete",
        "acknowledged": 0,
        "samples": samples,
        "telemetry_enabled": recorder is not None,
    }
    began = time.perf_counter_ns()
    try:
        with kind(os.environ[binding["runtime_dsn_envs"][0]]) as engine, sde.Session(
            model, placement, {name: engine}, recorder=recorder, project_id=config["project_id"]
        ) as session:
            if args.mode == "write":
                # Warm the same table/operation using a separate synthetic key namespace.
                for offset in range(2):
                    rows = [
                        reading(98, offset * args.batch_size + n + 1)
                        for n in range(args.batch_size)
                    ]
                    if args.method == "save":
                        session.save("WeatherReading", rows[0])
                    else:
                        session.save_many("WeatherReading", rows)
                began = time.perf_counter_ns()
                for first in range(1, args.rows + 1, args.batch_size):
                    rows = [
                        reading(1, n)
                        for n in range(first, min(first + args.batch_size, args.rows + 1))
                    ]
                    start = time.perf_counter_ns()
                    if args.method == "save":
                        session.save("WeatherReading", rows[0])
                    else:
                        session.save_many("WeatherReading", rows)
                    duration = time.perf_counter_ns() - start
                    samples.append(
                        {"operation": args.method, "duration_ns": duration, "rows": len(rows)}
                    )
                    report["acknowledged"] += len(rows)
            else:
                where = {"station": reading(99, 1)["station"]}
                cents = (
                    1525 * args.rows
                    + (args.rows // 1000) * 499500
                    + (args.rows % 1000) * (args.rows % 1000 + 1) // 2
                )
                for _ in range(20):
                    assert session.get(
                        "WeatherReading",
                        {
                            "station": where["station"],
                            "at": BASE_TIME + timedelta(microseconds=1),
                        },
                    ) == reading(99, 1)
                began = time.perf_counter_ns()
                for index in range(args.samples):
                    sequence = 1 + index * 104729 % args.rows
                    expected = reading(99, sequence)
                    result: Any
                    for operation in ("get", "scan", "count", "summarize"):
                        start = time.perf_counter_ns()
                        if operation == "get":
                            result = session.get(
                                "WeatherReading",
                                {"station": expected["station"], "at": expected["at"]},
                            )
                        elif operation == "scan":
                            result = session.scan(
                                "WeatherReading",
                                where=where,
                                bounds=sde.Range("at", low=expected["at"]),
                                limit=1000,
                            )
                        elif operation == "count":
                            result = session.count("WeatherReading", where=where)
                        else:
                            result = session.summarize("WeatherReading", "celsius", where=where)
                        duration = time.perf_counter_ns() - start
                        returned = 1
                        if operation == "get":
                            assert result == expected
                        elif operation == "scan":
                            returned = min(1000, args.rows - sequence + 1)
                            assert list(result.rows) == [
                                reading(99, sequence + n) for n in range(returned)
                            ]
                        elif operation == "count":
                            assert result == args.rows
                        else:
                            assert (
                                result.count == args.rows and result.non_null_count == args.rows
                            )
                            assert result.total == Decimal(cents).scaleb(-2)
                        samples.append(
                            {"operation": operation, "duration_ns": duration, "rows": returned}
                        )
            report["elapsed_ns"] = time.perf_counter_ns() - began
            report["status"] = "complete"
    finally:
        report["max_rss_kib"] = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
        if recorder:
            window = recorder.roll()
            assert window is not None
            report["window"] = window.as_record(model)
        args.output.write_text(json.dumps(report, allow_nan=False) + "\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
